third_party/move/move-vm/runtime/src/storage/ty_layout_converter.rs

```diff
@@ -185,229 +274,185 @@ pub(crate) trait LayoutConverterBase {
+                                },
+                            },
+                        };
+                        (layout, true)
+                    },
+                }
             },
-        })
-    }
-
-    fn struct_name_to_fully_annotated_layout(
-        &self,
-        struct_name_idx: StructNameIndex,
-        ty_args: &[Type],
-        count: &mut u64,
-        depth: u64,
-    ) -> PartialVMResult<MoveTypeLayout> {
-        let struct_type = self.fetch_struct_ty_by_idx(struct_name_idx)?;
-
-        // TODO(#13806): have annotated layouts for variants. Currently, we just return the raw
-        //   layout for them.
-        if matches!(struct_type.layout, StructLayout::Variants(_)) {
-            return self
-                .struct_name_to_type_layout(struct_name_idx, ty_args, count, depth)
-                .map(|(l, _)| l);
-        }
-
-        let struct_tag = self.struct_name_idx_to_struct_tag(struct_name_idx, ty_args)?;
-        let fields = struct_type.fields(None)?;
-
-        let field_layouts = fields
-            .iter()
-            .map(|(n, ty)| {
-                let ty = self
-                    .vm_config()
-                    .ty_builder
-                    .create_ty_with_subst(ty, ty_args)?;
-                let l = self.type_to_fully_annotated_layout_impl(&ty, count, depth)?;
-                Ok(MoveFieldLayout::new(n.clone(), l))
-            })
-            .collect::<PartialVMResult<Vec<_>>>()?;
-        let struct_layout =
-            MoveTypeLayout::Struct(MoveStructLayout::with_types(struct_tag, field_layouts));
-
-        Ok(struct_layout)
-    }
-}
-
-// --------------------------------------------------------------------------------------------
-// Layout converter based on ModuleStorage
-
-pub(crate) struct StorageLayoutConverter<'a> {
-    storage: &'a dyn ModuleStorage,
-}
+        };
 
-impl<'a> StorageLayoutConverter<'a> {
-    pub(crate) fn new(storage: &'a dyn ModuleStorage) -> Self {
-        Self { storage }
+        Ok(result)
     }
 }
 
-impl LayoutConverterBase for StorageLayoutConverter<'_> {
-    fn vm_config(&self) -> &VMConfig {
-        self.storage.runtime_environment().vm_config()
-    }
-
-    fn fetch_struct_ty_by_idx(&self, idx: StructNameIndex) -> PartialVMResult<Arc<StructType>> {
-        self.storage.fetch_struct_ty_by_idx(&idx)
-    }
-
-    fn struct_name_index_map(&self) -> &StructNameIndexMap {
-        self.storage.runtime_environment().struct_name_index_map()
-    }
-
-    fn struct_name_idx_to_struct_tag(
-        &self,
-        idx: StructNameIndex,
-        ty_args: &[Type],
-    ) -> PartialVMResult<StructTag> {
-        let ty_tag_builder = TypeTagConverter::new(self.storage.runtime_environment());
-        ty_tag_builder.struct_name_idx_to_struct_tag(&idx, ty_args)
-    }
+#[cfg(test)]
+mod tests {
+    // TODO(lazy-loading): add more tests here when all PRs are connected.
```

Comment:
Some things like delayed field substitutions should be well covered by more e2e/integration tests, but I do think it's critical to have good unit tests for the methods here - they can also act as an easy way for the reader to grasp the usage and subtle differences (which are exactly what it makes sense to unit test) between some of those APIs.

Reply:
+1, layouts are used a lot so we have pretty good coverage, my main concern are negative tests (bounds + cycles). For bounds we have 1-2 tests  somewhere in `e2e-move-tests` but it is relatively easy to have a unit test for this now that limits are part of the VM config. 

For cycles, there is a TODO in pr description, because what if:

```rust
struct A { b: B },
```
```rust
struct B { c: C },
```
```rust
struct C { a: A },
```

And then you call:
```
exists<A>(addr);
```

technically this is fine because of layout bounds on size, but we should be detecting these (a good question if this should be allowed, cause having recursive data types is pretty common). In any case, I want this to be more or less consistent with depth checks and not allowing cycles for now.

Reply:
Will add tests before landing 🫡